src/ir.cpp

```diff
@@ -23420,6 +23452,82 @@ static IrInstruction *ir_analyze_instruction_align_of(IrAnalyze *ira, IrInstruct
     return result;
 }
 
+static IrInstruction *ir_analyze_instruction_expect(IrAnalyze *ira, IrInstructionExpect *instruction) {
+    IrInstruction *value = instruction->value->child;
+    if (type_is_invalid(value->value.type))
+        return ira->codegen->invalid_instruction;
+
+    ZigType *value_type = value->value.type;
+
+    IrInstruction *expect_instr = instruction->expected->child;
+    if (type_is_invalid(expect_instr->value.type))
+        return ira->codegen->invalid_instruction;
+
+    switch (value_type->id) {
```

Comment:
I don't think this switch is needed at all. Implicit cast the expected value to the correct type, resolve it as comptime, and then either return `value` (in the case `value` is comptime known) or return an expect instruction.